RecoParticleFlow/PFClusterProducer/src/PFClusterEMEnergyCorrector.cc

```diff
@@ -7,33 +13,265 @@ namespace {
+            break;
+          case PFLayer::PS2:
+            ePS2 += psclus->energy();
+            break;
+          default:
+            break;
+          }
+        }
       }
+      
+      double correctedEnergy = _calibrator->energyEm(cluster,ePS1,ePS2,_applyCrackCorrections);
+      cluster.setCorrectedEnergy(correctedEnergy);
+      
     }
+    return;
+  }
+  
+  int bunchspacing = 450;  
+  
+  if (autoDetectBunchSpacing_) {
+    if (evt.isRealData()) {
+      edm::RunNumber_t run = evt.run();
+      if (run == 178003 ||
+          run == 178004 ||
+          run == 209089 ||
+          run == 209106 ||
+          run == 209109 ||
+          run == 209146 ||
+          run == 209148 ||
+          run == 209151) {
+        bunchspacing = 25;
+      }
+      else {
+        bunchspacing = 50;
+      }
+    }
+    else {
+      edm::Handle<int> bunchSpacingH;
+      evt.getByToken(bunchSpacing_,bunchSpacingH);
+      bunchspacing = *bunchSpacingH;
+    }
+  }
+  
+  const unsigned int ncor = 6;
+  
+  std::vector<edm::ESHandle<GBRForestD> > forestH_mean(ncor);
+  std::vector<edm::ESHandle<GBRForestD> > forestH_sigma(ncor);
+  
+  const std::vector<std::string> condnames_mean = (bunchspacing == 25) ? _condnames_mean_25ns : _condnames_mean_50ns;
+  const std::vector<std::string> condnames_sigma = (bunchspacing == 25) ? _condnames_sigma_25ns : _condnames_sigma_50ns;
+  
+  for (unsigned int icor=0; icor<ncor; ++icor) {
+    es.get<GBRDWrapperRcd>().get(condnames_mean[icor],forestH_mean[icor]);
+    es.get<GBRDWrapperRcd>().get(condnames_sigma[icor],forestH_sigma[icor]);
   }
-  const double eCorr= _calibrator->energyEm(cluster,
-					    ps1_energies,ps2_energies,
-					    ePS1,ePS2,
-					    _applyCrackCorrections);
-  cluster.setCorrectedEnergy(eCorr);
+  
+  std::array<float,11> eval;
+    
+  EcalClusterLazyTools lazyTool(evt, es, _recHitsEB, _recHitsEE);
+
+  //count number of primary vertices for pileup correction
+  edm::Handle<reco::VertexCollection> vtxH;
+  evt.getByToken(_vertices,vtxH);
+  int nvtx = 0;
+  for (const reco::Vertex &vtx : *vtxH) {
+    if (!vtx.isFake()) {
+      ++nvtx;  
+    }
+  }
+  
+  //magic numbers for MINUIT-like transformation of BDT output onto limited range
+  //(These should be stored inside the conditions object in the future as well)
+  const double meanlimlow = 1./1.4;
+  const double meanlimhigh = 1./0.4;
+  const double meanoffset = meanlimlow + 0.5*(meanlimhigh-meanlimlow);
+  const double meanscale = 0.5*(meanlimhigh-meanlimlow);
+  
+  const double sigmalimlow = 0.003;
+  const double sigmalimhigh = 0.5;
+  const double sigmaoffset = sigmalimlow + 0.5*(sigmalimhigh-sigmalimlow);
+  const double sigmascale = 0.5*(sigmalimhigh-sigmalimlow);  
+  
+  for (unsigned int idx = 0; idx<cs.size(); ++idx) {
+    
+    reco::PFCluster &cluster = cs[idx];
+    
+    double e = cluster.energy();
+    double eta = cluster.eta();
+    double phi = cluster.phi();    
+    
+    double invE = 1./e;
```

Comment:
It would be safer to guard against dividing by 0, in case some bug in the future might allow reco::PFCluster::energy() to return a 0 value. You could calculate invE this way:

if (e == 0.0)
   invE = 0.0;
else invE = 1.0/e;